File: TP01/q02/Palindromo.c
#include <stdio.h>
#include <string.h>

// função para verificar se uma string é um palíndromo
int verificarPalindromo(char texto[]) {
    int inicio = 0;
    int fim = strlen(texto) - 1;

    while (inicio < fim) {
        /* 
        a tabela ASCII só tem letras simples mas o UTF-8 inclui letras com acento e símbolos especiais:
        0xxx xxxx (0-127): caracteres ASCII padrão
        110x xxxx 10xx xxxx (128-2047): caracteres de 2 bytes
        caso o caractere seja uma letra especial, pula para o próximo
        */
        while ((texto[inicio] & 0xC0) == 0x80) 
            inicio++;
        while ((texto[fim] & 0xC0) == 0x80) 
            fim--;
        if (texto[inicio] != texto[fim]) {
            return 0; //se forem diferentes retorna falso
        }
        inicio++;
        fim--;
    }
    return 1; //se a palavra for um palíndromo retorna verdadeiro
}

int main() {
    char texto[500];

    //"flag" de continuação para o loop
    int continuar = 1;
    while (continuar && fgets(texto, 500, stdin)) { //enquanto continuar e houver linhas de entrada
        int tamanho = strlen(texto);

        if (tamanho > 0 && texto[tamanho - 1] == '\n') { //se a ultima posição for quebra de linha, substitui por "null"
            texto[tamanho - 1] = '\0';
        }
        
        //se a string for "FIM" troca o estado de continuar para 0
        if (strcmp(texto, "FIM") == 0) {
            continuar = 0;
        } else {
            int ehPalindromo = verificarPalindromo(texto);
            if (ehPalindromo) {
                printf("SIM\n");
            } else {
                printf("NAO\n");
            }
        }
    }
    

    return 0;
}

File: TP01/q01/Palindromo.java
package q01;

import java.util.*;

public class Palindromo {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in); //scanner para ler a entrada
        String texto;
        
        // continua lendo enquanto houver linhas e a entrada não for "FIM"
        while (scanner.hasNextLine() && !(texto = scanner.nextLine()).equals("FIM")) {
            boolean ehPalindromo = verificarPalindromo(texto);
            
            if (ehPalindromo) { //imprime a resposta
                System.out.println("SIM");
            } else {
                System.out.println("NAO");
            }
        }
        
        scanner.close();
    }
    
    public static boolean verificarPalindromo(String texto) {
        int inicio = 0;//inicializa os índices
        int fim = texto.length() - 1;
        
        while (inicio < fim) {
            if (texto.charAt(inicio) != texto.charAt(fim)) {//compara os caracteres
                return false;// se forem diferentes retorna falso
            }
            inicio++;//se forem iguais anda com os índices
            fim--;
        }
        
        return true;//se a palavra for de fato um palíndromo retorna verdadeiro
    }
}


